Sources/TuistGenerator/Utils/NamespaceGenerator.swift

```diff
@@ -0,0 +1,84 @@
+import PathKit
+import StencilSwiftKit
+import SwiftGenKit
+import TSCBasic
+import TuistSupport
+
+enum NamespaceType {
+    case assets
+    case strings
+
+    var templateFileName: String {
+        switch self {
+        case .assets:
+            return "xcassets.stencil"
+        case .strings:
+            return "strings.stencil"
+        }
+    }
+}
+
+private extension NamespaceType {
+    func parser() throws -> Parser {
+        switch self {
+        case .assets:
+            return try AssetsCatalog.Parser()
+        case .strings:
+            return try Strings.Parser()
+        }
+    }
+}
+
+protocol NamespaceGenerating {
+    func render(_ namespaceType: NamespaceType, paths: [AbsolutePath]) throws -> [(name: String, contents: String)]
+    func generateNamespaceScript() -> String
+}
+
+final class NamespaceGenerator: NamespaceGenerating {
+    private let resourcesNamespaceTemplatesLocator: ResourcesNamespaceTemplatesLocating
+
+    init(
+        resourcesNamespaceTemplatesLocator: ResourcesNamespaceTemplatesLocating = ResourcesNamespaceTemplatesLocator()
+    ) {
+        self.resourcesNamespaceTemplatesLocator = resourcesNamespaceTemplatesLocator
+    }
+
+    func render(_ namespaceType: NamespaceType, paths: [AbsolutePath]) throws -> [(name: String, contents: String)] {
+        let templatePath = try resourcesNamespaceTemplatesLocator.locateTemplate(for: namespaceType)
+        let template = StencilSwiftTemplate(
+            templateString: try FileHandler.shared.readTextFile(templatePath),
+            environment: stencilSwiftEnvironment()
+        )
+
+        return try paths.map { path in
+            let parser = try namespaceType.parser()
+            try parser.parse(path: Path(path.pathString), relativeTo: Path(""))
+            var context = parser.stencilContext()
+            context = try StencilContext.enrich(context: context, parameters: ["publicAccess": true])
+            return (path.basenameWithoutExt, try template.render(context))
+        }
+    }
+
+    func generateNamespaceScript() -> String {
+        #if DEBUG
+            // Used only for debug purposes to find currently-built tuist
+            // `bundlePath` points to .build/debug/tuist
+            let tuistCommand = AbsolutePath(#file.replacingOccurrences(of: "file://", with: ""))
+                .removingLastComponent()
+                .removingLastComponent()
+                .removingLastComponent()
+                .removingLastComponent()
+                .appending(components: ".build", "debug", "tuist")
+                .pathString
+        #else
+            let tuistCommand = "tuist"
+        #endif
+        return """
+        #!/bin/sh
+
+        pushd "${SRCROOT}"
+        \(tuistCommand) generate namespace
+        popd
+        """
```

Comment:
Worth noting, we previously followed a similar approach with `tuist embed` and we ended up steering away from it due to creating a strong coupling between the generated project and `tuist` - i.e. the generated project becomes will not build at all unless you have tuist installed. Additionally it means, users of `TuistGenerator` now rely on the `tuist` binary too :/ 

xref: https://github.com/tuist/tuist/issues/276

A few things we can consider:

- Control if the build phases are generated or not (more options, more confusion I suppose, perhaps only at the mapper level?)
- Can we vendor a small script / binary that is self contained into the `Derrived` directory? (I suspect it this may be quite a bit of hassle)
- Not generate a build phase at all, seeing that running `tuist generate` will produce the correct project each time along with the generated namespace and source, it's may be ok to expect users to run `tuist generate` if they add new resources. For example, that is the case today if you add a new Asset Catalog, or any resource, you'll need to re-generate the project

Reply:
That's a fair point that I have not realized - I think vendoring a small binary might not be such an hassle and as it will be added to `Derived` dir upon generation, the project will be buildable even without tuist. Let me try that and we can discuss it further if it doesn't work out 🙂

Reply:
I'm more leaning towards not adding the build phase at all. The scenario we'd be supporting with this is when the user adds a new resource by dragging and dropping the resource into the Xcode project. I think in that case developers could simply run a command from the terminal, or generate the whole project.
The reason why I lean more towards not generating assets at build time is because the caching feature is not aware of what's happening through build phases and that means the feature might not work as expected if we go down this path.

Reply:
I have not thought about it from the caching point-of-view - I suppose I then agree with you, although, it will need some getting used to since I am used to just build the project to regenerate the related Swift code after adding asset/changing strings. This might be a good time to try to [push this](https://github.com/tuist/tuist/issues/183). A co-worker of mine, @svastven, has been looking into it, so hopefully we can add it soon enough 🙂